### PR TITLE
refactor: sdk improvements

### PR DESCRIPTION
**Pull request by ctian1**

### CODE REVIEW

File: prover/src/lib.rs [hunk 1]
```diff
@@ -458,9 +481,13 @@ impl SP1Prover {
     pub fn wrap_bn254(
         &self,
         vk: &SP1VerifyingKey,
-        core_challenger: Challenger<CoreSC>,
         reduced_proof: SP1ReduceProof<InnerSC>,
     ) -> ShardProof<OuterSC> {
+        // Get verify_start_challenger from the reduce proof's public values.
+        let pv = RecursionPublicValues::from_vec(reduced_proof.proof.public_values.clone());
```
[the remark this answers is not in the document]
Comment (ctian1) [reply prover/src/lib.rs:487]: no, it's checked in the program when `is_complete=1` is committed, so we can just check that

File: prover/src/lib.rs [hunk 2]
```diff
@@ -478,11 +505,49 @@ impl SP1Prover {
     }
 
     /// Wrap the STARK proven over a SNARK-friendly field into a Groth16 proof.
-    pub fn wrap_groth16(&self, proof: ShardProof<OuterSC>) {
+    pub fn wrap_groth16(&self, proof: ShardProof<OuterSC>, build_dir: PathBuf) -> Groth16Proof {
+        let pv = RecursionPublicValues::from_vec(proof.public_values.clone());
+
+        // Convert pv.vkey_digest to a bn254 field element
+        let mut vkey_hash = Bn254Fr::zero();
+        for (i, word) in pv.sp1_vk_digest.iter().enumerate() {
```
[the remark this answers is not in the document]
Comment (ctian1) [reply prover/src/lib.rs:640]: this is not stable yet, it's not constrained in circuit

File: prover/src/lib.rs [hunk 3]
```diff
@@ -103,19 +108,44 @@ impl SP1Prover {
     /// Creates a proving key and a verifying key for a given RISC-V ELF.
     pub fn setup(&self, elf: &[u8]) -> (SP1ProvingKey, SP1VerifyingKey) {
         let program = Program::from(elf);
-        let config = CoreSC::default();
-        let machine = RiscvAir::machine(config);
-        let (pk, vk) = machine.setup(&program);
+        let (pk, vk) = self.core_machine.setup(&program);
```
[the remark this answers is not in the document]
Comment (ctian1) [reply prover/src/lib.rs:154]: yea program is in SP1ProvingKey already actually

**Comment by matthiasgoergens** [comment]
This should probably have been squash-merged, or its history been cleaned up before a merge.

**Pull request merged**